oauth_provider/models/oauth_provider_scope.py

```diff
@@ -0,0 +1,338 @@
+                on.
+
+        Returns:
+            OauthProviderScope: Recordsets associated to the model.
+        """
+        if isinstance(model, models.BaseModel):
+            model = model._name
+        return self.filtered(lambda record: record.model == model)
+
+    @api.multi
+    def get_data(self, model_name, record_ids=None, all_scopes_match=False,
+                 domain=None):
+        """ Return the data matching the scopes from the requested model.
+
+        Args:
+            model_name (str): Name of the model to operate on.
+            record_ids (list[int]): ID of record(s) to find. Will only return
+                this record, if defined.
+            all_scopes_match (bool): True to filter out records that do not
+                match all of the scopes in the current recordset.
+            domain (list of tuples, optional): Domain to append to the
+                `filter_domain` that is defined in the scope.
+
+        Returns:
+            dict:
+                This will be a dictionary of scoped record data, keyed by
+                record ID.
+                If `record_ids` is defined, and only one ID, the inner data
+                dictionary will be returned without nesting.
+        """
+
+        data = defaultdict(dict)
+        eval_context = self.ir_filter_eval_context
+        all_scopes_records = self.env[model_name]
+        record_ids = record_ids or []
+
+        for scope in self.filter_by_model(model_name):
+
+            records = scope._get_scoped_records(
+                model_name, eval_context, domain, record_ids,
+            )
+
+            for record_data in records.read(scope.field_ids.mapped('name')):
+                for field, value in record_data.items():
+                    if isinstance(value, tuple):
+                        # Return only the name for a many2one
+                        data[record_data['id']][field] = value[1]
+                    else:
+                        data[record_data['id']][field] = value
+
+            # Keep a list of records that match all scopes
+            if not all_scopes_records:
+                all_scopes_records = records
+            else:
+                all_scopes_records &= records
+
+        # If all scopes are required to match, filter the results to keep only
+        # those matching all scopes
+        if all_scopes_match:
+            data = dict(
+                filter(
+                    lambda _data: _data[0] in all_scopes_records.ids,
+                    data.items(),
+                ),
+            )
+
+        matched_records = {
+            rec_id: data.get(rec_id, {}) for rec_id in record_ids
+        }
+        if matched_records:
+            if len(matched_records) == 1:
+                return matched_records.values()[0]
+            return matched_records
+
+        return data
+
+    @api.multi
+    def create_record(self, model_name, vals):
```

Comment:
I don't understand why this method is in the `oauth.provider.scope` model, instead of calling the three methods (`_validate_scope_field`, `self.env['model_name'].create` and `_validate_scope_record`) directly in the `create_record` method of `oauth.provider.token` ?
What's the benefit of having this method on the scope ?

Same question about write and delete, obviously.